javascript/javascript1/week4/homework/voice-assistant.js

```diff
@@ -0,0 +1,185 @@
+        return;
+    }
+    for(i=0; i < namesList.length; i++){
+        if (nameOfUser == namesList[i]){
+            console.log("The name is already used!");
+            return;
+        }
+    }
+    namesList.push(nameOfUser);
+    console.log("Nice to meet you " + nameOfUser);
+}
+
+function replyWithLastSavedName(){
+    if (namesList.length == 0) {
+        console.log("You haven't told me your name yet!");  
+        return;
+    }
+    else {
+        let lastNameOfArrey = namesList[namesList.length-1]
+        console.log("Your name is " + lastNameOfArrey);
+        return;
+    }
+}
+
+const todosList = [];
+function addToTodoList(command){
+    let firstIndex = "Add".length;
+    let secondIndex = command.indexOf(" to my todo");
+    let newToDo = command.slice(firstIndex, secondIndex).trim();
+    todosList.push(newToDo);
+    console.log(newToDo + " added to your todo");
+}
+
+
+
+function removeFromTodoList(command){
+    let firstIndex = "Remove".length;
+    let secondIndex = command.indexOf(" from my todo");
+    let fromToDo = command.slice(firstIndex, secondIndex).trim();
+    remove(todosList, fromToDo);
+}
+
+function remove(todosList, fromToDo) {
+    let indexToRemove = todosList.indexOf(fromToDo);
+    while (indexToRemove != -1) {
+        todosList.splice(indexToRemove, 1);
+        indexToRemove = todosList.indexOf(fromToDo);
+    }
+    console.log("Removed " + fromToDo + " from your todo");
+}
+
+function getTodoList(){
+    if (todosList == null){
+        console.log(" you have no todos!");
+        return;
+    }
+    console.log(`you have ${todosList.length} todos- ${todosList.join("and")}`);
+}
+
+const months = [
+    'January',
+    'February',
+    'March',
+    'April',
+    'May',
+    'June',
+    'July',
+    'August',
+    'September',
+    'October',
+    'November',
+    'December'
+];
+const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
+
+function getTodayFormattedDate() {
+    const today = new Date();
+    const year = today.getFullYear();
+    const date = today.getDate();
+    const monthIndex = today.getMonth();
+    const monthName = months[monthIndex];
+    const dayIndex = today.getDay();
+    const dayName = days[dayIndex];
+    const formattedDate = `${dayName}, ${date}. of ${monthName} ${year}`
+    console.log(formattedDate);
+}
+
+function doSimpleMath(command){
+    const simpleMath = command.slice(7).trim();
+    if(validMathExpression(simpleMath)){ // For security reasons we should validate first
+        try{
+            const calcResult = new Function("return " + simpleMath + ";")();
+            if(isNaN(calcResult)){
+                console.log("Invalid matth problem!");
+            }
+            else{
+                console.log("It is: ", calcResult);
+            }
+        }
+        catch(e){
+            console.log("Invalid matth problem!");
+        }
+    }
+}
+
+const rg = /^[0-9\+\-\*\/\s]+$/;
```

Comment:
I don't think your regular expression is correct.  The best way to solve this is to try with a couple of examples. 
Let's take  expression " 1 + 5 ". this can be split in "1" , " " , "+" and "5".
Therefore you can match with a number [0-9] , \s (space) , character '+', \s(space) and [0-9]
so you could have a patter like this `patt = /[0-9]\s\+\s[0-9]`. This would work with the example above...but if you want to include other operators then you could do something like bellow
```
  var str = "1999 + 249";
  var patt1 = /[0-9]+\s(\*|\+|-|:)\s[0-9]+/g;
  console.log(str.match(patt1)) 
```
My feedback is to start with simple exemple first. try to make that work...For me would be sufficient if your code would have just work with the addition and with expressions like ' 1 + 5'

Reply:
Also then you can rely on ` var result2  = str.split('+');` and you would get an array with [1999,249]. There are solutions that can work with any expression like you wrote (  '3 + 3 + 3') but it requires more work...and it's not needed here

Reply:
Thanks for the input, the explanation of the problem was not very clear and sort of ambiguous so I wrote it in a way that it covers any arithmetic input. As I read, the downside for using eval or Function was security concerns so I used the regular expression barely to make sure that there is nothing in the string except numbers and arithmetic operators that can cause any problem. I'm not yet on top of regular expressions though and I think I should spend some days in order to master it and use it similar to the one you have provided above. However, I have fully tested mine and it's working and it makes the code much simpler than extracting each number and operator from the text.
I have committed some changes though to make sure it covers also some invalid input edge cases by adding a try-catch and controlling the output result.
As I've learned regular expression is quite a complicated subject and is taught as a major part of a semester course in universities. I think it deserved to be addressed in our classes.